value/structreflect.go

```diff
@@ -0,0 +1,238 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package value
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"sync/atomic"
+)
+
+// reflectStructCache keeps track of json tag related data for structs and fields to speed up reflection.
+// TODO: This overlaps in functionality with the fieldCache in
+// https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/converter.go#L57 but
+// is more efficient at lookup by json field name. The logic should be consolidated. Only one copy of the cache needs
+// to be kept for each running process.
+var (
+	reflectStructCache = newStructCache()
+)
+
+type structCache struct {
+	// use an atomic and copy-on-write since there are a fixed (typically very small) number of structs compiled into any
+	// go program using this cache
+	value atomic.Value
+	// mu is held by writers when performing load/modify/store operations on the cache, readers do not need to hold a
+	// read-lock since the atomic value is always read-only
+	mu sync.Mutex
+}
+
+type structCacheMap map[reflect.Type]structCacheEntry
+
+// structCacheEntry contains information about each struct field, keyed by json field name, that is expensive to
+// compute using reflection.
+type structCacheEntry map[string]*fieldCacheEntry
+
+// Get returns true and fieldCacheEntry for the given type if the type is in the cache. Otherwise Get returns false.
+func (c *structCache) Get(t reflect.Type) (map[string]*fieldCacheEntry, bool) {
+	entry, ok := c.value.Load().(structCacheMap)[t]
+	return entry, ok
+}
+
+// Update sets the fieldCacheEntry for the given type via a copy-on-write update to the struct cache.
+func (c *structCache) Update(t reflect.Type, m map[string]*fieldCacheEntry) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
```

Comment:
Possible optimization: check if `t` is already present (due to multiple simultaneous computations when we lose the race) and bail, since the build function is idempotent (right?). Might make a difference on startup.

Reply:
Nice, yes, that might help, and certainly won't hurt. The map entry puts are idempotent. Added.